### PR TITLE
Add backup currencies rate sources

### PR DESCRIPTION
**Pull request by andiserg**

### CODE REVIEW

File: internal/adapters/currencyrate/readers/currency_api.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	}
+}
+
+func (cr *CurrencyAPICurrencyReader) GetCurrencyRate(from string, to string) (float32, error) {
+	data, err := ReadHTTP(cr.APIURL + "&base_currency=" + strings.ToUpper(from))
+	if err != nil {
+		return 0, err
+	}
+
+	if rates, ok := (*data)["data"].(map[string]interface{}); ok {
```
Comment (andriihar) [inline internal/adapters/currencyrate/readers/currency_api.go:29]: Can we use struct here?

File: internal/adapters/currencyrate/readers/fawaza_api.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+func (cr *FawazaAPICurrencyReader) GetCurrencyRate(from string, to string) (float32, error) {
+	from = strings.ToLower(from)
+	to = strings.ToLower(to)
+
+	data, err := ReadHTTP(cr.APIURL + strings.ToLower(from) + ".json")
+	if err != nil {
+		return 0, err
+	}
+
+	if rates, ok := (*data)[from].(map[string]interface{}); ok {
```
Comment (andriihar) [inline internal/adapters/currencyrate/readers/fawaza_api.go:32]: And here?

File: internal/adapters/currencyrate/factory.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package currencyrate
+
+import (
+	readers2 "go_service/internal/adapters/currencyrate/readers"
```
Comment (andriihar) [inline internal/adapters/currencyrate/factory.go:4]:
```suggestion
	readers "go_service/internal/adapters/currencyrate/readers"
```

File: internal/app/mailingTask.go [hunk 4]
```diff
@@ -68,7 +68,7 @@ func (ms RateMailer) RunSending() {
 
 func (ms RateMailer) SendRateToAll() error {
 	subscribers := ms.subscriberGateway.GetAll()
-	rate, err := ms.currencyGateway.GetUSDCurrencyRate()
+	rate, err := ms.currencyGateway.GetCurrencyRate("", "")
```
Comment (andriihar) [inline internal/app/mailingTask.go:71]: Is it ok with empty strings?

File: internal/adapters/currencyrate/readers/currency_api.go (hunk 1 above)
Comment (andiserg) [reply internal/adapters/currencyrate/readers/currency_api.go:29]: We can't use structs here, because a response's body is not unified

**Comment by andiserg** [comment]
The comments were resolved in the next Pull Request #12

File: internal/adapters/currencyrate/readers/currency_api.go (hunk 1 above)
Comment (andriihar) [reply internal/adapters/currencyrate/readers/currency_api.go:29]: But, than I didn`t understand the logic.

You have three independent providers. All of them have method `GetCurrencyRate` and can encapsulate `getRates` with specific struct in them, didn't you?


image

**Pull request merged**